### PR TITLE
Support multiple operations per ACLRule

### PR DESCRIPTION
**Pull request by OuesFa**
Signed-off-by: fares.oueslati <oueslati.fares@gmail.com>

### Type of change

- Enhancement / new feature

### Description

Solves https://github.com/strimzi/strimzi-kafka-operator/issues/1671

Currently I need to add a new ACLRule for every operation I want to permit. The KafkaUser CRD is getting very long this way when the user needs read and write access to a lot of topics.

Adding an `operations` and deprecating `operation`

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [ ] Make sure all tests pass
- [x] Update documentation
- [] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md
- [] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: api/src/main/java/io/strimzi/api/kafka/model/AclRule.java [hunk 1]
```diff
@@ -45,6 +49,14 @@ public AclRule(AclRuleType type, AclRuleResource resource, String host, AclOpera
...
     }
 
+    public AclRule(AclRuleType type, AclRuleResource resource, String host, List<AclOperation> operations) {
+        this.type = type;
+        this.resource = resource;
+        this.host = host;
+        this.operations = operations;
+    }
+
+
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/AclRule.java:59]: Why do we need both this and the previous constructor without the `List<AclOperation> operations` field?

File: api/src/test/resources/io/strimzi/api/kafka/model/KafkaUser.yaml [hunk 2]
```diff
@@ -13,18 +13,18 @@ spec:
           type: topic
           name: my-topic
           patternType: prefix
-        operation: Read
+        operations: ["Read","Write"]
```
Comment (scholzj) [inline api/src/test/resources/io/strimzi/api/kafka/model/KafkaUser.yaml:16]: Here and elsewhere ... are the quotes really needed? I don' think they should be. Also, maybe we should use here an encoding as list?

```yaml
operations:
- Read
- Write
```

File: api/src/test/resources/io/strimzi/api/kafka/model/KafkaUser.yaml [hunk 3]
```diff
@@ -13,18 +13,18 @@ spec:
...
           name: my
           patternType: prefix
-        operation: Read
+        operations: ["Read","Describe","Write","Create"]
         host: 127.0.0.1
       - resource:
           type: group
           name: my-group
-        operation: Read
+        operations: ["Read"]
```
Comment (scholzj) [inline api/src/test/resources/io/strimzi/api/kafka/model/KafkaUser.yaml:26]: I think you should have a test which tests the `operation` field as well. It is still there so it should be tested.

File: user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java [hunk 4]
```diff
@@ -133,13 +136,22 @@ public static SimpleAclRule fromAclBinding(AclBinding aclBinding) {
...
+    @SuppressWarnings("deprecation")
+    public static List<SimpleAclRule> fromCrd(AclRule rule) {
+        if (rule.getOperations() != null) {
+            List<SimpleAclRule> simpleAclRules = new ArrayList<>();
+            for (AclOperation operation : rule.getOperations()) {
+                simpleAclRules.add(new SimpleAclRule(rule.getType(), SimpleAclRuleResource.fromCrd(rule.getResource()), rule.getHost(), operation));
+            }
+            return simpleAclRules;
+        } else {
+            return Collections.singletonList(new SimpleAclRule(rule.getType(), SimpleAclRuleResource.fromCrd(rule.getResource()), rule.getHost(), rule.getOperation()));
```
Comment (scholzj) [inline user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java:153]: Maybe you can use `List.of(...)` instead of `Collections.singletonList(...)`.

File: user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java [hunk 5]
```diff
@@ -133,13 +136,22 @@ public static SimpleAclRule fromAclBinding(AclBinding aclBinding) {
...
+        }
```
Comment (scholzj) [inline user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java:154]: You should probably throw some error (`InvalidResourceException`) in case both are set.

File: api/src/main/java/io/strimzi/api/kafka/model/AclRule.java [hunk 6]
```diff
@@ -82,7 +94,8 @@ public void setHost(String host) {
 
     @Description("Operation which will be allowed or denied. " +
             "Supported operations are: Read, Write, Create, Delete, Alter, Describe, ClusterAction, AlterConfigs, DescribeConfigs, IdempotentWrite and All.")
-    @JsonProperty(required = true)
+    @DeprecatedProperty(description = "This property is replaced by the `operations` property")
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/AclRule.java:97]: The deprecated property annotation has fields `movedToPath` which you should use. It will generate the docs for the move.

File: user-operator/src/test/java/io/strimzi/operator/user/model/acl/SimpleAclRuleTest.java [hunk 7]
```diff
@@ -46,11 +50,30 @@ public void testFromCrd()   {
...
-        assertThat(simple.getResource(), is(resource));
+        List<SimpleAclRule> simpleAclRules = SimpleAclRule.fromCrd(rule);
+        assertThat(simpleAclRules.get(0).getOperation(), is(AclOperation.READ));
+        assertThat(simpleAclRules.get(0).getType(), is(AclRuleType.ALLOW));
+        assertThat(simpleAclRules.get(0).getHost(), is("127.0.0.1"));
+        assertThat(simpleAclRules.get(0).getResource(), is(resource));
+    }
+
+    @Test
+    public void testFromCrdWithThreeOperationsPerRule()   {
```
Comment (scholzj) [inline user-operator/src/test/java/io/strimzi/operator/user/model/acl/SimpleAclRuleTest.java:61]: Are there some tests which still test the old layout with single operation?

**Review by scholzj — COMMENTED** [review]
Thanks for the PR. I added some nits. Also, you need to run `make crd_install` to copy the CRDs into the Helm Chart as well.

File: CHANGELOG.md [hunk 8]
```diff
@@ -10,6 +10,7 @@
   * Kafka Connect, Mirror Maker, Mirror Maker 2 and Strimzi Bridge can be configured to use OpenTelemetry
   * Using Jaeger exporter by default for backward compatibility
 * Updated JMX Exporter dependency to 0.17.2
+* Support multiple operations per ACLRule
```
Comment (scholzj) [inline CHANGELOG.md:13]: Maybe you can also add to the subchapter below a note about the deprecation of the single operation field.

File: api/src/main/java/io/strimzi/api/kafka/model/AclRule.java (hunk 1 above)
Comment (OuesFa) [reply api/src/main/java/io/strimzi/api/kafka/model/AclRule.java:59]: It doesn't seem to be used anymore, indeed 👍

File: user-operator/src/test/java/io/strimzi/operator/user/model/acl/SimpleAclRuleTest.java (hunk 7 above)
Comment (OuesFa) [reply user-operator/src/test/java/io/strimzi/operator/user/model/acl/SimpleAclRuleTest.java:61]: I think the test above does it.

**Comment by scholzj** [comment]
Looks like the build failed because of some checkstyle issues:

```
[ERROR] /home/vsts/work/1/s/user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java:8:34: Using the '.*' form of import should be avoided - io.strimzi.api.kafka.model.*. [AvoidStarImport]
[ERROR] /home/vsts/work/1/s/user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java:9:41: Using the '.*' form of import should be avoided - io.strimzi.operator.cluster.model.*. [AvoidStarImport]
[ERROR] /home/vsts/work/1/s/user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java:10:35: Using the '.*' form of import should be avoided - org.apache.kafka.common.acl.*. [AvoidStarImport]
[ERROR] /home/vsts/work/1/s/user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java:11:40: Using the '.*' form of import should be avoided - org.apache.kafka.common.resource.*. [AvoidStarImport]
[ERROR] /home/vsts/work/1/s/user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java:12:45: Using the '.*' form of import should be avoided - org.apache.kafka.common.security.auth.*. [AvoidStarImport]
[ERROR] /home/vsts/work/1/s/user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java:14:17: Using the '.*' form of import should be avoided - java.util.*. [AvoidStarImport]
[ERROR] /home/vsts/work/1/s/user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java:148:48: '{' is not preceded with whitespace. [WhitespaceAround]
```

File: api/src/test/java/io/strimzi/api/kafka/model/KafkaUserCrdIT.java [hunk 9]
```diff
@@ -49,7 +49,7 @@ void testKafkaUserWithExtraProperty() {
             KubeClusterException.class,
             () -> createDeleteCustomResource("KafkaUser-with-extra-property.yaml"));
 
-        assertThat(exception.getMessage(), containsString("unknown field \"thisPropertyIsNotInTheSchema\""));
+        assertThat(exception.getMessage(), containsString("unknown field \"spec.thisPropertyIsNotInTheSchema\""));
```
Comment (scholzj) [inline api/src/test/java/io/strimzi/api/kafka/model/KafkaUserCrdIT.java:52]: Same as above. This seems quite strange as I would not expect this to change. What version of Kubernetes did you run the tests against?

File: api/src/test/resources/io/strimzi/api/kafka/model/JmxTrans-output-definition-with-missing-required-property.yaml [hunk 10]
```diff
@@ -9,9 +9,6 @@ spec:
     storage:
       type: persistent-claim
       size: 500Gi
-    listeners:
-      plain: {}
-      tls: {}
```
Comment (scholzj) [inline api/src/test/resources/io/strimzi/api/kafka/model/JmxTrans-output-definition-with-missing-required-property.yaml:14]: I think these changes might be valid But it would be best to leave them for a separate PR to make the reviews of this one easier and more straight forward.

**Comment by OuesFa** [comment]
> 

Oh ok thanks I didn't pay attention to the checkstyle because at the moment there are several tests that fail locally even though I haven't modified the classes they cover so it seems odd, but I'll fix them anyway and push with the checkstyle.

File: development-docs/DEV_GUIDE.md [hunk 11]
```diff
@@ -78,6 +78,9 @@ structure after everything is said and done should look like `/Library/Java/Java
 doing that run the command at the beginning again and this should link the file and allow you to use maven with OpenJDK
 version 11.
 
+When running the tests, you may encounter `openssl` related errors for parts that you may not have even worked on, in 
+which case please consider upgrading your `openssl` version.
+
```
Comment (scholzj) [inline development-docs/DEV_GUIDE.md:83]: It might be good to try to figure this out since nobody else reported any issues. What version of OpenSSL are you using?

Comment (OuesFa) [reply development-docs/DEV_GUIDE.md:83]: I was using 2.8.3 when I had the issues.
Now that I'm using the OpenSSL 3.0.5 it works fine.

File: api/src/test/java/io/strimzi/api/kafka/model/KafkaUserCrdIT.java (hunk 9 above)
Comment (OuesFa) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaUserCrdIT.java:52]: I'm using v1.25.0 to test.

File: api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java [hunk 12]
```diff
@@ -51,7 +51,7 @@ void testKafkaMirrorMaker2WithMissingRequired() {
             KubeClusterException.class,
             () -> createDeleteCustomResource("KafkaMirrorMaker2-with-missing-required-property.yaml"));
 
-        assertMissingRequiredPropertiesMessage(exception.getMessage(), "connectCluster", "clusters.alias", "sourceCluster", "targetCluster");
+        assertMissingRequiredPropertiesMessage(exception.getMessage(), "spec.connectCluster", "spec.clusters[2].alias", "sourceCluster", "targetCluster");
```
[the remark this answers is not in the document]
Comment (OuesFa) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java:54]: I'm using v1.25.0 to test.

File: api/src/test/resources/io/strimzi/api/kafka/model/JmxTrans-output-definition-with-missing-required-property.yaml (hunk 10 above)
Comment (OuesFa) [reply api/src/test/resources/io/strimzi/api/kafka/model/JmxTrans-output-definition-with-missing-required-property.yaml:14]: It's because of the testing issues I had, it might be because of the Kubernetes version I'm running the tests against though.

File: api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (OuesFa) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java:54]: Should I change the Kube version to avoid those errors? 🤔

[the remark this answers is not in the document]
Comment (scholzj) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java:54]: I'm running these against 1.25 as well. And I do not have any issues like that. They are all passing on the CIs as well. So that seems weird - like something in your environment is braking them. What is your `kubectl` version as some of these use `kubectl`

[the remark this answers is not in the document]
Comment (OuesFa) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java:54]:
```
kubectl version                                                                                                                                                                                                                   

WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.0", GitCommit:"a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2", GitTreeState:"clean", BuildDate:"2022-08-23T17:44:59Z", GoVersion:"go1.19", Compiler:"gc", Platform:"darwin/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.0", GitCommit:"a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2", GitTreeState:"clean", BuildDate:"2022-08-23T17:38:15Z", GoVersion:"go1.19", Compiler:"gc", Platform:"linux/amd64"}
```

[the remark this answers is not in the document]
Comment (scholzj) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java:54]: Turns out I had an old `kubectl`. With Kubectl 1.25 it fails for me as well. Sorry. That said, I do not think the fixes aou did will work on the CI which uses kubectl / Kubernetes 1.19. It would be also good to separate it from this PR because it will be much easier to review but also to make sure the tests work locally but also on the CI. 

In the cases where things changed, you will need probably to use the `anyOf` instead of just replacing the existing values. If you want, I can look at it and open the PR. Or I can leave it to you. As you prefer.

[the remark this answers is not in the document]
Comment (OuesFa) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java:54]: Ok that's clear, yes I will do the change if possible.
I'm just not sure to understand the following 

> In the cases where things changed, you will need probably to use the anyOf instead of just replacing the existing values

I can see that `anyOf` [is already used here](https://github.com/strimzi/strimzi-kafka-operator/blob/5bb792252b92338e9f4f0951d55d322be0b62769/api/src/test/java/io/strimzi/api/kafka/model/AbstractCrdIT.java#L116)
You mean add other `containsStringIgnoringCase` statements to match the new patterns?

[the remark this answers is not in the document]
Comment (OuesFa) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java:54]: Yes, makes sense, I'm on it. I will push a fix asap. 
Do I need to do it in a separate branch/PR or on this one?

File: api/src/test/resources/io/strimzi/api/kafka/model/JmxTrans-output-definition-with-missing-required-property.yaml (hunk 10 above)
Comment (OuesFa) [reply api/src/test/resources/io/strimzi/api/kafka/model/JmxTrans-output-definition-with-missing-required-property.yaml:14]: I'm fixing the tests by adding patterns to the `anyOf` list of matchers, however, for this one it's a bit tricky 

Here is the error I get if I keep the `listeneres` property
```
`kubectl --namespace kafkacrd-it create -f /Users/f.oueslati/repos/strimzi-kafka-operator/api/target/test-classes/io/strimzi/api/kafka/model/JmxTrans-output-definition-with-missing-required-property.yaml` got status code 1 and stderr:
------
Error from server (BadRequest): error when creating "/Users/f.oueslati/repos/strimzi-kafka-operator/api/target/test-classes/io/strimzi/api/kafka/model/JmxTrans-output-definition-with-missing-required-property.yaml": Kafka in version "v1beta2" cannot be handled as a Kafka: strict decoding error: unknown field "spec.kafka.listeners.plain", unknown field "spec.kafka.listeners.tls" 
```

So, I won't be able to match anything on `outputType` or `name` [as it is expected in the test](https://github.com/strimzi/strimzi-kafka-operator/blob/main/api/src/test/java/io/strimzi/api/kafka/model/KafkaCrdIT.java#L152) 

Should I change the test behaviour to match against `unknown field "spec.kafka.listeners.plain"`? By renaming the test to `testJmxOptionsWithUnknownFields` or should I keep the changes above? Or maybe something else?

File: api/src/test/resources/io/strimzi/api/kafka/model/Kafka-with-missing-required-property.yaml [hunk 13]
```diff
@@ -3,7 +3,9 @@ kind: Kafka
 metadata:
   name: missing-required-property
 spec:
-  topicOperator:
```
Comment (OuesFa) [inline api/src/test/resources/io/strimzi/api/kafka/model/Kafka-with-missing-required-property.yaml:6]: Same here, if I don't fix the yaml, the test won't pass with the expected behavior.

File: api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (scholzj) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java:54]: I think that separate branch / PR will be better, because it would make it easier to test and review. For example, you will see if it passes in the CI without being affected by the ACL changes etc.

File: api/src/test/resources/io/strimzi/api/kafka/model/Kafka-with-missing-required-property.yaml (hunk 13 above)
Comment (scholzj) [reply api/src/test/resources/io/strimzi/api/kafka/model/Kafka-with-missing-required-property.yaml:6]: Yeah, I think you need to fix the YAML here.

File: api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (OuesFa) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java:54]: Indeed!

[the remark this answers is not in the document]
Comment (OuesFa) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java:54]: Here it is :)

[the remark this answers is not in the document]
Comment (OuesFa) [reply api/src/test/java/io/strimzi/api/kafka/model/KafkaMirrorMaker2CrdIT.java:54]: I forgot to put the link sorry https://github.com/strimzi/strimzi-kafka-operator/pull/7433

File: development-docs/DEV_GUIDE.md (hunk 11 above)
Comment (scholzj) [reply development-docs/DEV_GUIDE.md:83]: I don't think that openSSL has any verion 2.8. Are you on Mac? In that case, it might have been LibreSSL which MacOS defaults to. So maybe we can be more specific here and say that you should make sure you are using OpenSSL and not LibreSSL?

File: user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java [hunk 14]
```diff
@@ -133,13 +138,24 @@ public static SimpleAclRule fromAclBinding(AclBinding aclBinding) {
...
+        if (rule.getOperations() != null) {
+            List<SimpleAclRule> simpleAclRules = new ArrayList<>();
+            for (AclOperation operation : rule.getOperations()) {
+                simpleAclRules.add(new SimpleAclRule(rule.getType(), SimpleAclRuleResource.fromCrd(rule.getResource()), rule.getHost(), operation));
+            }
+            return simpleAclRules;
+        } else if (rule.getOperation() != null) {
+            return List.of(new SimpleAclRule(rule.getType(), SimpleAclRuleResource.fromCrd(rule.getResource()), rule.getHost(), rule.getOperation()));
+        } else {
+            throw new InvalidResourceException("Both fields `operations` and `operation` cannot be filled in at the same time");
```
Comment (scholzj) [inline user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java:157]: I do not think this will work => It they will be both set, it will fall into the first section and use the `operations` field.

**Comment by OuesFa** [comment]
Seems to be better now :)

**Review by scholzj — COMMENTED** [review]
Some more comments:
* In the `api` module, we should have a test which decodes / encodes the old format with `operation`. It seems you changed the existing tests to use 
`operations` which is fine. But we should have this covered as well.

But it looks good otherwise.

**Comment by OuesFa** [comment]
> Some more comments:
> 
> * In the `api` module, we should have a test which decodes / encodes the old format with `operation`. It seems you changed the existing tests to use
>   `operations` which is fine. But we should have this covered as well.
> 
> But it looks good otherwise.

Do you think I should put `operations` and `operation` in the same `api/src/test/resources/io/strimzi/api/kafka/model/KafkaUser.out.yaml` file? Or should I create another file to test only `operation` in it?

File: user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java (hunk 14 above)
Comment (OuesFa) [reply user-operator/src/main/java/io/strimzi/operator/user/model/acl/SimpleAclRule.java]: I also added a simple test about this behaviour.

**Review by scholzj — APPROVED** [review]
Thanks a lot for the PR. LGTM.

**Comment by scholzj** [comment]
> EDIT: I used KafkaUserV1beta1.yaml to test the deprecated operation field.

Yeah, I think that is completely sufficient.

**Comment by OuesFa** [comment]
> Thanks a lot for the PR. LGTM.

Thank you for your guidance and the amazing job you're doing!

**Review by ppatierno — APPROVED** [review]
